app/controllers/credentials/credentials-form.js

```diff
@@ -0,0 +1,49 @@
+function formatErrorsForSummaryList (errors = {}) {
+  return Object.entries(errors).map(([id, message]) => ({
+    href: `#${id}`,
+    text: message
+  }))
+}
+class CredentialsForm {
+  constructor (fields = []) {
+    this.fields = fields
+    this.values = {}
+  }
+
+  from (entity = {}) {
+    const values = this.fields.reduce((aggregate, field) => {
+      const key = field.key || field.id
+      aggregate[field.id] = entity[key]
+      return aggregate
+    }, {})
+    return { values }
+  }
+
+  validate (formData = {}) {
+    const errors = {}
+    const values = {}
+
+    this.fields.forEach((field) => {
+      const valid = field.valid || []
+      values[field.id] = typeof formData[field.id] === 'string' ? formData[field.id].trim() : formData[field.id]
+      valid.some((validator) => {
+        const valid = validator.method(formData[field.id])
+        if (!valid) {
+          errors[field.id] = validator.message
+          return true
+        }
+      })
+    })
+    return {
+      values,
+      errors,
+      errorSummaryList: formatErrorsForSummaryList(errors)
+    }
+  }
+}
+
+function isNotEmpty (value) {
```

Comment:
Even when we're sharing this code for any form, does it make more sense for this function live in a separate validation helper class, which would maybe replace `server-side-form-validations.js` when we more widely use this pattern?